dvc/repo/brancher.py

```diff
@@ -0,0 +1,53 @@
+def brancher(  # noqa: E302
+    self, branches=None, all_branches=False, tags=None, all_tags=False
+):
+    """Generator that iterates over specified revisions.
+
+    Args:
+        branches (list): a list of branches to iterate over.
+        all_branches (bool): iterate over all available branches.
+        tags (list): a list of tags to iterate over.
+        all_tags (bool): iterate over all available tags.
+
+    Yields:
+        str: the current revision.
+    """
+    if not any([branches, all_branches, tags, all_tags]):
+        yield "Working Tree"
+        return
+
+    saved_tree = self.tree
+    revs = []
+
+    scm = self.scm
+
+    if self.scm.is_dirty():
+        from dvc.repo.tree import WorkingTree
+
+        self.tree = WorkingTree()
+        yield "Working Tree"
```

Comment:
This breaks tests that expect previous behaviour of yielding "" if it is not from any branch. Could we continue doing that?

Reply:
Oops. Not this one actually, but the same thing in line 16. I did run the tests many times, but not after the line 16 change which I added at the last moment :-/.

I think it is better to clarify in the docstring that what the brancher yeilds is not a branch names (instead it is just the abstract name of the currently selected tree), and adjust the tests for this behavior. What do you think? @efiop @shcheklein

Reply:
It is used to display the name for metrics in different branches.

Reply:
Oh... and in `dvc push` output, where it is reduntant for sure...

Reply:
So I think it is still ok to leave "Working Tree" string for iteration with `--all-branches`, but yield an empty string when there is no things to iterate over.

Reply:
The other solution could be to not use the `brancher()` if there were no `--all-branches` / `--all-tags` specified. But I'd prefer to stick to the current one.

Reply:
Maybe it could make sense to replace the "Working Tree" with the "index" string.

Reply:
@ei-grad "index" is even more confusing IMHO :) Empty string feels nice, but maybe I'm just too used to it 🙂

Reply:
Updated the docstring to about what the brancher could yeild.

Reply:
so, this is about printing a tree name when we do some output, right? First, I would make a tree itself return the name, no need then to yield anything and pass it around, we will be able to access it from the repo, the same way we access the current object. Is it possible? Second, in the FsTree we can return "Working Tree" or "*" or "workspace", or whatever :) Btw, i'm more or less fine with the current solution as well, it's not ideal but it's not hard to fix anyway. As for the empty string. @efiop does it mean we will have to do something like `if not branch: name  "Working Tree"` further along, in multiple place? I would try to avoid that.

Reply:
@shcheklein No, I don't see why we even need "Working Tree" anywhere.

Reply:
In `dvc metrics show -a` when we print output we need a way to distinguish "workspace" from an actual branch. May be some other places, @ei-grad should know better than me by now :)